File: python/0229.py
# Link: https://leetcode.com/problems/majority-element-ii/description/
# Runtime: 102 ms
# Memory: 17.77 MB


class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        candidate1, candidate2, count1, count2 = 0, 1, 0, 0
        for val in nums:
            if val == candidate1:
                count1 += 1
            elif val == candidate2:
                count2 += 1
            elif count1 == 0:
                candidate1, count1 = val, 1
            elif count2 == 0:
                candidate2, count2 = val, 1
            else:
                count1 -= 1
                count2 -= 1

        return [
            num for num in (candidate1, candidate2) if nums.count(num) > len(nums) // 3
        ]

File: c++/0229.cpp
// Link: https://leetcode.com/problems/majority-element-ii/description/
// Runtime: 7 ms
// Memory: 16.23 MB

class Solution {
public:
  std::vector<int> majorityElement(const std::vector<int> &nums) {
    int candidate1 = 0, candidate2 = 1, count1 = 0, count2 = 0;
    std::vector<int> majorityElements;
    for (const auto &val : nums) {
      if (val == candidate1)
        count1++;
      else if (val == candidate2)
        count2++;
      else if (count1 == 0) {
        candidate1 = val;
        count1 = 1;
      } else if (count2 == 0) {
        candidate2 = val;
        count2 = 1;
      } else {
        count1--;
        count2--;
      }
    }

    if (std::count(nums.begin(), nums.end(), candidate1) > nums.size() / 3.0)
      majorityElements.push_back(candidate1);
    if (std::count(nums.begin(), nums.end(), candidate2) > nums.size() / 3.0)
      majorityElements.push_back(candidate2);

    return majorityElements;
  }
};
